### ISSUE TITLE
Mac: python cannot load xrootd client libs

### ISSUE THREAD

**Issue by Axel-Naumann** [xrootd/xrootd#959]
We get
```
cling::DynamicLibraryManager::loadLibrary(): dlopen(/Volumes/HDD2/ec/build/night/LABEL/mac1012/SPEC/noimt/build/lib/libNetxNG.so, 9): Library not loaded: libXrdUtils.2.dylib
  Referenced from: /Volumes/HDD2/ec/build/night/LABEL/mac1012/SPEC/noimt/build/lib/libNetxNG.so
  Reason: unsafe use of relative rpath libXrdUtils.2.dylib in /Volumes/HDD2/ec/build/night/LABEL/mac1012/SPEC/noimt/build/lib/libNetxNG.so with restricted binary
Error in <TInterpreter::TCling::AutoLoad>: failure loading library libNetxNG.so for TNetXNGSystem
```
on our Mac builds. That's caused by python loading a library (libNetxNG.so from ROOT) which depends on `libXrdUtils.2.dylib`. `otool` says:
```
otool -L lib/libNetxNG.so
lib/libNetxNG.so:
	@rpath/libNetxNG.so (compatibility version 0.0.0, current version 0.0.0)
	libXrdUtils.2.dylib (compatibility version 2.0.0, current version 2.0.0)
	libXrdCl.2.dylib (compatibility version 2.0.0, current version 2.0.0)
	@rpath/libNet.so (compatibility version 0.0.0, current version 0.0.0)
	@rpath/libRIO.so (compatibility version 0.0.0, current version 0.0.0)
	@rpath/libThread.so (compatibility version 0.0.0, current version 0.0.0)
	@rpath/libCore.so (compatibility version 0.0.0, current version 0.0.0)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 400.9.4)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1252.250.1)
```

As you can see, the xrootd client libs don't have an RPATH set. I see no way of forcing one upon them; without, binaries in SIP-protected areas (like `/usr/bin/python`) refuse to load the xrootd client libs.

Could RPATH be allowed as an option, or am I misunderstanding something here?

Cheers, Axel.

**Comment by Axel-Naumann** [comment]
You can reproduce this e.g. with ROOT's xrootd library using
```
> python
Python 2.7.10 (default, Feb 22 2019, 21:17:52)
[GCC 4.2.1 Compatible Apple LLVM 10.0.1 (clang-1001.0.37.14)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from ctypes import *
>>> cdll.LoadLibrary("lib/libNetxNG.so")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/ctypes/__init__.py", line 443, in LoadLibrary
    return self._dlltype(name)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/ctypes/__init__.py", line 365, in __init__
    self._handle = _dlopen(self._name, mode)
OSError: dlopen(lib/libNetxNG.so, 6): Library not loaded: libXrdUtils.2.dylib
  Referenced from: /Users/axel/build/root/cmake/lib/libNetxNG.so
  Reason: unsafe use of relative rpath libXrdUtils.2.dylib in lib/libNetxNG.so with restricted binary
>>>
```

**Comment by simonmichal** [comment]
@Axel-Naumann : thanks for reporting this problem and for the reproducer, I will certainly add an rpath option to our cmake infrastructure.

Michal

**Comment by simonmichal** [comment]
@Axel-Naumann : I just pushed 80d5fd629b12675ce33c59b1acce0e136049da7d and then refined with 28af21e69eb95eee09dd8af11f0ef1ce82b64c2b ;-) it enables `MACOSX_RPATH` by default. You can also overwrite this by using the `-DCMAKE_MACOSX_RPATH` if you would happen to wish to do so. This should solve your problem.

BTW when building root with xrootd_builtin flag I noticed that you guys are building the whole xrootd project, wouldn't it be sufficient to build just the client? 
I'm asking because for a month now we are building only the client on MACOSX with a assumption nobody actually needs the server side on mac. Let me know your thoughts!

**Comment by Axel-Naumann** [comment]
Awesome, thanks for fixing and even more (!) for going the extra mile and finding a bug in our build procedure, much appreciated! Would you be able to point me to the right flag to pass to CMake to build client-only?

**Comment by simonmichal** [comment]
Any time :-)

In order to build just the client simply run make from `build/src/XrdCl` directory.

**Comment by simonmichal** [comment]
@Axel-Naumann : do you have any feedback for us? is it OK if I close this issue?

**Comment by simonmichal** [comment]
Closing it, if there is need to follow up in the future don't hesitate to reopen!

**Comment by Axel-Naumann** [comment]
Sorry, I'm slow.

I cannot compile https://github.com/xrootd/xrootd/commit/28af21e69eb95eee09dd8af11f0ef1ce82b64c2b because:
```
/Users/axel/build/root/cmake/XROOTD-prefix/src/XROOTD/src/XrdOfs/XrdOfs.cc:636:63: error: use of undeclared identifier 'O_DIRECT'
       if ((error.getUCap() & fRedir) == fRedir) open_flag |= O_DIRECT;
                                                              ^
1 error generated.
```

**Comment by simonmichal** [comment]
@Axel-Naumann : this has been fixed in ab1fa57dca1facc5e7b7674f65237ff7caa86040, though I still encourage you to build only the client side ;-)

**Comment by Axel-Naumann** [comment]
Confirmed, f7982b9ce82714405fd06d5dd0eeeef6dc430876 works, thanks!
